File: JavaToKotlin/apps(kotlin)/src/main/kotlin/io/dustin/apps/board/api/usecase/community/comment/ReadCommentUseCase.java
package io.dustin.apps.board.api.usecase.community.comment;

import io.dustin.apps.board.domain.community.comment.model.dto.CommentDto;
import io.dustin.apps.board.domain.community.comment.service.ReadCommentService;
import io.dustin.apps.common.model.CountByPagingInfo;
import io.dustin.apps.common.model.QueryPage;
import io.dustin.apps.common.model.ResponseWithScroll;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

import static io.dustin.apps.common.model.ResponseWithScrollSetting.getCountByPagingInfo;

@Service
@RequiredArgsConstructor
public class ReadCommentUseCase {

    private final ReadCommentService readCommentService;

    public ResponseWithScroll<List<CommentDto>> execute(Long postingId, QueryPage queryPage) {
        /**
         * 게시물에 대한 순수 댓글 리스트만 보여줌
         */
        int realSize = queryPage.getSize();
        int querySize = realSize + 1;
        long userId = 1;
        List<CommentDto> result = readCommentService.getCommentsByPosting(userId, postingId, querySize, queryPage.getNextId());
        CountByPagingInfo<CommentDto> cbi = getCountByPagingInfo(result, realSize);

        return ResponseWithScroll.from(cbi.result(), cbi.isLast(), cbi.nextId());

    }

    public ResponseWithScroll<List<CommentDto>> replyListByComment(Long commentId, QueryPage queryPage) {
        /**
         * 대댓글 여부 판단하기
         */
        int realSize = queryPage.getSize();
        int querySize = realSize + 1;
        long userId = 1;
        List<CommentDto> result = readCommentService.replyListByComment(userId, commentId, querySize, queryPage.getNextId());
        CountByPagingInfo<CommentDto> cbi = getCountByPagingInfo(result, realSize);

        return ResponseWithScroll.from(cbi.result(), cbi.isLast(), cbi.nextId());

    }


}

File: apps/src/main/kotlin/io/dustin/apps/board/api/usecase/community/comment/ReadCommentUseCase.kt
package io.dustin.apps.board.api.usecase.community.comment

import ResponseWithScrollSetting.getCountByPagingInfo
import io.dustin.apps.board.api.community.request.query.AllCommentQuery
import io.dustin.apps.board.api.community.request.query.CommentDetailQuery
import io.dustin.apps.board.api.community.request.query.PostingDetailQuery
import io.dustin.apps.board.domain.community.comment.model.dto.CommentDto
import io.dustin.apps.board.domain.community.comment.service.ReadCommentService
import io.dustin.apps.common.model.CountByPagingInfo
import io.dustin.apps.common.model.QueryPage
import io.dustin.apps.common.model.ResponseWithScroll
import io.dustin.apps.common.model.response.ResultResponsePagination
import io.dustin.apps.common.utils.lastAt
import org.springframework.stereotype.Service

@Service
class ReadCommentUseCase (

    private val readCommentService: ReadCommentService

) {
    fun execute(query: PostingDetailQuery): ResultResponsePagination<CommentDto> {

        val userId: Long = 1

        val list = readCommentService.getCommentsByPosting(userId, query.postingId, query.recordsCount, query.nextId)
        println("몇개나 들어있을까요" + list.size)
        if(list.isEmpty()) {
            return ResultResponsePagination.of(
                last = true,
                recordsCount = 0,
                data = list,
            )
        }
        val result = list.take((query.recordsCount).toInt())

        val isLast = list.size <= query.recordsCount.toInt()

        val nextId = if (!isLast) {
            result.lastAt().id
        } else {
            null
        }

        return ResultResponsePagination.of(
            nextId = nextId,
            last = isLast,
            recordsCount = result.size.toLong(),
            data = result,
        )

    }

    fun replyListByComment(query: CommentDetailQuery): ResultResponsePagination<CommentDto> {

        val userId: Long = 1

        val list: List<CommentDto> = readCommentService.replyListByComment(userId, query.commentId, query.recordsCount, query.nextId)
        println("몇개나 들어있을까요" + list.size)
        if(list.isEmpty()) {
            return ResultResponsePagination.of(
                last = true,
                recordsCount = 0,
                data = list,
            )
        }
        val result = list.take((query.recordsCount).toInt())

        val isLast = list.size <= query.recordsCount.toInt()

        val nextId = if (!isLast) {
            result.lastAt().id
        } else {
            null
        }

        return ResultResponsePagination.of(
            nextId = nextId,
            last = isLast,
            recordsCount = result.size.toLong(),
            data = result,
        )
    }


}